### PR TITLE
feat(frontend): support execute insert in local mode

### PR DESCRIPTION
**Pull request by ZENOTME**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

related issue: #7684

support to execute insert without using select query in local mode, such as: `insert into t values (1)`

For other dml like: insert-select, delete, update, their plan has more than two stage so that we can't execute them in local mode.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

- SQL commands, functions, and operators

### Release note

</details>

In past the insert will be executed in distributed mode, but now the insert without using select query will executed in local mode.

### CODE REVIEW

File: src/frontend/src/optimizer/mod.rs [hunk 1]
```diff
@@ -576,11 +601,7 @@ impl PlanRoot {
             ctx.trace("To Batch Distributed Plan:");
             ctx.trace(plan.explain_to_string().unwrap());
         }
-        if has_batch_insert(plan.clone())
-            || has_batch_delete(plan.clone())
-            || has_batch_update(plan.clone())
-            || Self::require_additional_exchange_on_root(plan.clone())
-        {
+        if Self::require_additional_exchange_on_root(plan.clone()) {
```
Comment (ZENOTME) [inline src/frontend/src/optimizer/mod.rs:604]: `require_additional_exchange_on_root` now will find the insert, delete, update no exchange before so I think we can remove has_** here. 

Although the behaviour of `has_**` and `exist_and_no_exchange_before` is different, I think it's still correct now? CMIIW.

**Comment by lmatz** [comment]
Do you want to test `insert` in both local mode and distributed mode in the nightly performance test?
https://buildkite.com/risingwave-test/sysbench

**Comment by ZENOTME** [comment]
> Do you want to test `insert` in both local mode and distributed mode in the nightly performance test? https://buildkite.com/risingwave-test/sysbench

Yes

**Comment by ZENOTME** [comment]
> Do you want to test `insert` in both local mode and distributed mode in the nightly performance test? https://buildkite.com/risingwave-test/sysbench

Yes, how I test that

**Comment by lmatz** [comment]
> > Do you want to test `insert` in both local mode and distributed mode in the nightly performance test? https://buildkite.com/risingwave-test/sysbench
> 
> Yes, how I test that

QA team will add local mode into the existing performance test pipeline

File: src/frontend/src/handler/query.rs [hunk 2]
```diff
@@ -43,12 +43,42 @@ use crate::scheduler::{
...
+    ) | is_insert_using_select(stmt))
+}
+
 pub fn gen_batch_query_plan(
     session: &SessionImpl,
     context: OptimizerContextRef,
     stmt: Statement,
 ) -> Result<(PlanRef, QueryMode, Schema)> {
-    let stmt_type = to_statement_type(&stmt)?;
+    let must_dist = must_run_in_distributed_mode(&stmt)?;
```
Comment (kwannoel) [inline src/frontend/src/handler/query.rs:81]: Should this also check for QUERY_MODE session variable set to `distributed`?

File: src/frontend/src/scheduler/local.rs [hunk 3]
```diff
@@ -453,4 +454,39 @@ impl LocalQueryExecution {
...
+                            .worker_node_manager()
+                            .get_workers_by_parallel_unit_ids(&parallel_unit_ids)?;
+                        candidates.choose(&mut rand::thread_rng()).unwrap().clone()
+                    };
+                    Ok(vec![worker_node])
+            } else {
+                Ok(vec![self.front_env.worker_node_manager().next_random()?])
+            }
+        } else {
+            Ok(self.front_env.worker_node_manager().list_worker_nodes())
```
Comment (kwannoel) [inline src/frontend/src/scheduler/local.rs:489]: Hmm should we list all worker nodes here?
What if parallelism = 3, and worker nodes = 5, does that mean we choose 5 workers?
Should it match exactly?

File: src/frontend/src/scheduler/local.rs [hunk 4]
```diff
@@ -453,4 +454,39 @@ impl LocalQueryExecution {
...
+        if stage.parallelism.unwrap() == 1 {
+            if let NodeBody::Insert(insert_node) = &stage.root.node
+                && let Some(vnode_mapping) = self.get_vnode_mapping(&insert_node.table_id.into()) {
+                    let worker_node = {
+                        let parallel_unit_ids = vnode_mapping.iter_unique().collect_vec();
+                        let candidates = self.front_env
+                            .worker_node_manager()
+                            .get_workers_by_parallel_unit_ids(&parallel_unit_ids)?;
+                        candidates.choose(&mut rand::thread_rng()).unwrap().clone()
+                    };
```
Comment (kwannoel) [inline src/frontend/src/scheduler/local.rs:483]: Why do we need `get_workers_by_parallel_unit_ids` here? Didn't use this before, so not too familiar.
In the case of `Insert` why is this needed?

File: src/frontend/src/optimizer/mod.rs [hunk 5]
```diff
@@ -504,6 +472,76 @@ fn inline_session_timezone_in_exprs(ctx: OptimizerContextRef, plan: PlanRef) ->
...
+    fn is_delete(plan: &PlanRef) -> bool {
+        plan.node_type() == PlanNodeType::BatchDelete
+    }
+
+    assert_eq!(plan.distribution(), &Distribution::Single);
+    exist_and_no_exchange_before(&plan, is_user_table)
+        || exist_and_no_exchange_before(&plan, is_source)
+        || exist_and_no_exchange_before(&plan, is_insert)
+        || exist_and_no_exchange_before(&plan, is_update)
+        || exist_and_no_exchange_before(&plan, is_delete)
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply src/frontend/src/optimizer/mod.rs:519]: Oh I see, they won't be executed in local because requires two stage.

File: src/frontend/src/scheduler/local.rs (hunk 3 above)
Comment (liurenjie1024) [reply src/frontend/src/scheduler/local.rs:489]: I think it should be workers with target table only.

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM, just require some refinement.

File: src/frontend/src/scheduler/local.rs (hunk 4 above)
Comment (ZENOTME) [reply src/frontend/src/scheduler/local.rs:483]: The insert executor will send the insert data to the reader registered by the dml executor.
https://github.com/risingwavelabs/risingwave/blob/1ad23babf8b61e89cbe9ffc3a8fadb7f878f9ffe/src/stream/src/executor/dml.rs#L94
Hence to access the reader, the insert executor need to schedule the same worker node of the dml executor. So `get_workers_by_parallel_unit_ids` is to get the worker node where dml executor stay in.

File: src/frontend/src/scheduler/local.rs (hunk 3 above)
Comment (ZENOTME) [reply src/frontend/src/scheduler/local.rs]: We have process case that the stage have scan node before, so in this case the stage don't have the scan node, I think we just need to randomlly select N parallelism worker?🤔

(Previous process is list worker node)
https://github.com/risingwavelabs/risingwave/blob/1ad23babf8b61e89cbe9ffc3a8fadb7f878f9ffe/src/frontend/src/scheduler/local.rs#L325

File: src/frontend/src/scheduler/local.rs (hunk 4 above)
Comment (kwannoel) [reply src/frontend/src/scheduler/local.rs:483]: I see, thanks for the clear explanation!

Maybe this can be documented, since the logic is split in various places, it does not seem very clear to me at first glance.
(Unless some documentation already exists, in that case feel free to ignore).

Comment (ZENOTME) [reply src/frontend/src/scheduler/local.rs:483]: Agree with you. (Seems don't have a related doc)

**Review by liurenjie1024 — APPROVED** [review]
LGTM
